pilot/pkg/proxy/envoy/v2/xds.go

```diff
@@ -24,75 +27,135 @@ import (
 	"google.golang.org/grpc"
 
 	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/v1alpha3"
 )
 
 // ConfigCache for xDS resources
 type ConfigCache struct {
+	serviceSuffix string
+
+	// inputs
 	services model.ServiceDiscovery
 	configs  model.ConfigStore
 
+	// version counter
+	// TODO: this should be derived from the inputs for load balancing between xDS instances to work correctly
+	version int64
+
+	// generated hosts for the version counter;
+	// hosts need to be computed once per the entire mesh, reducing the load on specializing it to each proxy
+	hosts []v1alpha3.GuardedHost
+
+	// tracking nodes with the last request timestamp
+	// TODO: garbage collection based on the expiration of the last request or stream closure
+	nodes map[string]time.Time
+
 	snapshots cache.SnapshotCache
 	server    server.Server
+
+	mu sync.Mutex
 }
 
 // NewConfigCache spins up a new config cache
-func NewConfigCache(services model.ServiceDiscovery, configs model.ConfigStore) *ConfigCache {
+func NewConfigCache(services model.ServiceDiscovery, configs model.ConfigStore, domainSuffix string) *ConfigCache {
 	out := &ConfigCache{
-		services: services,
-		configs:  configs,
+		serviceSuffix: "svc." + domainSuffix,
```

Comment:
???

Reply:
That's part of derivation of the short names. Where am I supposed to get the domain suffix? The API is badly misaligned with respect to k8s conventions, and I'm trying to do the best I can.